tests/lint-groups.rs

```diff
@@ -0,0 +1,132 @@
+//! This test checks and updates `book/src/granular_lint_groups.md`
+//!
+//! Specifically the parts in between lint-group-start and lint-group-end are not meant to be edited
+//! by hand and are instead generated based on the const `GROUPS` slice.
+#![feature(rustc_private)]
+
+use std::collections::HashSet;
+use std::{env, fs};
+
+use clippy_lints::LintInfo;
+use clippy_lints::declared_lints::LINTS;
+use indoc::formatdoc;
+use itertools::Itertools;
+use regex::{Captures, Regex};
+
+const GROUPS: &[(&str, &[&str])] = &[
+    ("perf-pedantic", &[
+        "assigning_clones",
+        "inefficient_to_string",
+        "naive_bytecount",
+        "needless_bitwise_bool",
+        "trivially_copy_pass_by_ref",
+        "unnecessary_join",
+        "unnecessary_box_returns",
+    ]),
+    ("perf-restriction", &[
+        "format_push_string",
+        "missing_asserts_for_indexing",
+    ]),
+    ("perf-nursery", &[
+        "redundant_clone",
+        "iter_with_drain",
+        "mutex_integer",
+        "or_fun_call",
+        "significant_drop_tightening",
+        "trivial_regex",
+        "needless_collect",
+    ]),
+    ("panicking", &[
+        "arithmetic_side_effects",
+        "expect_used",
+        "unwrap_used",
+        "panic",
+        "unreachable",
+        "todo",
+        "unimplemented",
+        "string_slice",
+        "indexing_slicing",
+    ]),
+    ("debugging", &["dbg_macro", "todo", "unimplemented"]),
+];
+
+#[test]
+fn check_lint_groups() {
+    let file = fs::read_to_string("book/src/granular_lint_groups.md").expect("failed to read granular_lint_groups.md");
+    let all_lint_names: HashSet<_> = LINTS
+        .iter()
+        .map(|LintInfo { lint, .. }| lint.name.strip_prefix("clippy::").unwrap().to_ascii_lowercase())
+        .collect();
+
+    let regex = Regex::new(
+        "(?s)\
+        (?<header><!-- lint-group-start: (?<name_start>[\\w-]+) -->)\
+        (?<lints>.*?)\
+        (?<footer><!-- lint-group-end: (?<name_end>[\\w-]+) -->)\
+        ",
+    )
```

Comment:
In `clippy_dev::update_lints` there is a function called `replace_region_in_{file,text}` that does the same thing as this, but without regex.

Why not implement this in `clippy_dev` and use the facilities that are already there. It is probably a good extension for the `update_lints` subcommand.

Reply:
Regex was slightly more convenient here because we also need to capture part of the start pattern, which `replace_region_in_text` didn't really allow and now needs small amounts of manual parsing, but I think it's fine.

This also required a bunch of changes to `replace_region_in_text`:
- The text in between the start/end pattern in `replace_region_in_text` is now passed to the replacement function since we need it to get the subgroup name
- It now does multiple replacements if needed (necessary here because there are multiple `<!-- lint-subgroup -->` sections)